Adds blog for SigV4 support in clients (#1217)
opensearch-project/project-website: _posts/2022-12-09-aws-sigv4-support-for-clients.md

---
layout: post
title:  "AWS SigV4 support for OpenSearch clients"
authors:
  - hvamsi
  - theotr
  - xtansia
  - wbeckler
  - monicank
date: 2022-12-09
categories:
  - feature

---

## Introduction

> OpenSearch Clients now support the ability to sign requests using [AWS Signature V4](https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html). This has been a community request since we forked clients from Elasticsearch, and we’re happy to announce that we have completed work across all clients, in collaboration with external contributors. Signing requests using native clients has been a big requirement for accessing the Amazon OpenSearch Service on AWS using fine grained access controls. Having SigV4 support natively in clients avoids using cURL requests or other workarounds.

Reviewer: ```suggestion
OpenSearch Clients now support the ability to sign requests using [AWS Signature V4](https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html). This has been a community request since we forked clients from Elasticsearch, and we’re happy to announce that we have completed work across all clients, in collaboration with external contributors. Signing requests using native clients has been an essential requirement for accessing the Amazon OpenSearch Service on AWS using fine grained access controls. Having SigV4 support natively in clients avoids using cURL requests and other workarounds.
```

## Setting up the managed service to use fine grained access control

> Be sure to update the access control to an IAM role/user. Do not use master user. Here the IAM role used allows access to this specific OpenSearch domain.

Reviewer: ```suggestion
Be sure to update the access control to an IAM role/user. Do not use master user. In the following image, the IAM role allows access to this specific OpenSearch domain:
```

![Fine Grained Access Control]({{ site.baseurl }}/assets/media/blog-images/2022-12-09-aws-sigv4-support-for-clients/fine-grained-control.png){: .img-fluid}

> You can even set a domain level access policy without using fine grained access. Ensure that the IAM role used has read/write access to the domain.

Reviewer: ```suggestion
Alternatively, you can set a domain level access policy without using fine grained access. Ensure that the IAM role you use has read/write access to the domain.
```

## Creating a client connection using SigV4 signing

 Before you start, ensure that you have [AWS credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html) set up on your machine. AWS credentials are stored in `~/.aws/credentials` and contain an access key and a secret key that allow you to authenticate with AWS resources using IAM. 

### Creating a client connection in Python

The python client requires you to have `boto3` installed. Make sure to update the `connection_class` to use `RequestsHttpConnection`. Use the following code to create a client connection in python:

```python
from opensearchpy import OpenSearch, RequestsHttpConnection, AWSV4SignerAuth
import boto3
host = '' # cluster endpoint, for example: my-test-domain.us-east-1.es.amazonaws.com
region = 'us-west-2'
credentials = boto3.Session().get_credentials()
auth = AWSV4SignerAuth(credentials, region)
client = OpenSearch(
    hosts = [{'host': host, 'port': 443}],
    http_auth = auth,
    use_ssl = True,
    verify_certs = True,
    connection_class = RequestsHttpConnection
)
```

### Creating a client connection in JavaScript

> The javascript client requires you to have `aws-sdk` installed. Depending on which version of the sdk you are currently using, initalize the client appropriately as given below:

Reviewer: ```suggestion
The javascript client requires you to have `aws-sdk` installed. Depending on which version of the sdk you are using, initalize the client appropriately as given below:
```

      // Any other method to acquire a new Credentials object can be used.
      const credentialsProvider = defaultProvider();
      return credentialsProvider();
    },
  }),
  node: "https://search-xxx.region.es.amazonaws.com", // OpenSearch domain URL
});

```
### Creating a client connection in Ruby

The [opensearch-aws-sigv4](https://github.com/opensearch-project/opensearch-ruby/tree/main/opensearch-aws-sigv4) gem provides the `OpenSearch::Aws::Sigv4Client` class, which has all features of `OpenSearch::Client`. The only difference between these two clients is that `OpenSearch::Aws::Sigv4Client` requires an instance of `Aws::Sigv4::Signer` during instantiation to authenticate with AWS:

```ruby
require 'opensearch-aws-sigv4'
require 'aws-sigv4'
signer = Aws::Sigv4::Signer.new(service: 'es',
                                region: 'us-west-2',
                                access_key_id: 'key_id',
                                secret_access_key: 'secret')
client = OpenSearch::Aws::Sigv4Client.new({ log: true }, signer)
client.cluster.health
client.transport.reload_connections!
client.search q: 'test'
```

### Creating a client connection in .NET

> All required request signing is handled by the `AwsSigV4HttpConnection` implementation, by default it will use the .NET AWS SDK's default credentials provider 

Reviewer: ```suggestion
All required request signing is handled by the `AwsSigV4HttpConnection` implementation. By default, `AwsSigV4HttpConnection` uses the .NET AWS SDK's default credentials provider 
```

> to acquire credentials from the environment. However, you may opt to pass in your own credentials provider; e.g. to assume a role. Refer to the [OpenSearch.Net User Guide](https://github.com/opensearch-project/opensearch-net/blob/main/USER_GUIDE.md#opensearchnetauthawssigv4) for complete getting started instructions.

Reviewer: ```suggestion
to acquire credentials from the environment. However, you may opt to pass in your own credentials provider, for example, to assume a role. Refer to the [OpenSearch.Net User Guide](https://github.com/opensearch-project/opensearch-net/blob/main/USER_GUIDE.md#opensearchnetauthawssigv4) for complete instructions.
```

```c#
using OpenSearch.Client;
using OpenSearch.Net.Auth.AwsSigV4;

var endpoint = new Uri("https://search-xxx.region.es.amazonaws.com");
var connection = new AwsSigV4HttpConnection();
var config = new ConnectionSettings(endpoint, connection);
var client = new OpenSearchClient(config);
```

### Creating a client connection in Rust

> Request signing is configured using the [`Credentials::AwsSigV4`](https://docs.rs/opensearch/latest/opensearch/auth/enum.Credentials.html#variant.AwsSigV4) enum variant or its helper conversion from an AWS SDK config. See [aws-config](https://docs.rs/aws-config/latest/aws_config/) for other AWS credentials provider implementations; e.g. assume role.

Reviewer: ```suggestion
Request signing is configured using the [`Credentials::AwsSigV4`](https://docs.rs/opensearch/latest/opensearch/auth/enum.Credentials.html#variant.AwsSigV4) enum variant or its helper conversion from an AWS SDK config. See [aws-config](https://docs.rs/aws-config/latest/aws_config/) for other AWS credentials provider implementations, for example, to assume a role.
```

```rust
use opensearch::{
    cat::CatIndicesParts,
    http::transport::{SingleNodeConnectionPool, TransportBuilder},
    OpenSearch,
};
use url::Url;

let creds = aws_config::load_from_env().await;

let host = "https://search-xxx.region.es.amazonaws.com";
let transport = TransportBuilder::new(SingleNodeConnectionPool::new(Url::parse(host).unwrap()))
    .auth(creds.try_into()?)
    // .auth(Credentials::AwsSigV4(creds.credentials().unwrap().clone(), creds.region().unwrap().clone()))
    .build()?;
let client = OpenSearch::new(transport);
```

## Conclusion

> You can now sign their requests using the client APIs natively instead of using workarounds. We’re continuing to work on improving the capabilities of SigV4 in clients with scenarios like async connections, compressed requests, and connection pooling support. We’re happy to take pull requests and feedback in the form of issues on [github](https://github.com/opensearch-project/opensearch-py/issues).

Reviewer: ```suggestion
You can now sign your requests using the client APIs natively instead of using workarounds. We’re continuing to work on improving the capabilities of SigV4 in clients with scenarios like async connections, compressed requests, and connection pooling support. We’re happy to take pull requests and feedback in the form of issues on [github](https://github.com/opensearch-project/opensearch-py/issues).
```

---
layout: post
title:  "AWS SigV4 support for OpenSearch clients"
authors:
  - vachshah
  - dblock
  - hvamsi
  - theotr
  - xtansia
  - mnkugler
  - mtimmerm
  - ssayakci
date: 2022-12-09
categories:
  - feature

---

## Introduction

OpenSearch Clients now support the ability to sign requests using [AWS Signature V4](https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html). This has been a community request since we forked clients from Elasticsearch, and we’re happy to announce that we have completed work across all clients, in collaboration with external contributors. Signing requests using native clients has been an essential requirement for accessing the Amazon OpenSearch Service on AWS using fine grained access controls. Having SigV4 support natively in clients avoids using cURL requests and other workarounds.

## Setting up the managed service to use fine grained access control

> Be sure to update the access control to an IAM role/user. Do not use master user. In the following image, the IAM role used allows access to this specific OpenSearch domain:

Reviewer: @harshavamsi Could you confirm that the new wording accurately reflects the meaning?

> Be sure to update the access control to an IAM role/user. Do not use master user. In the following image, the IAM role used allows access to this specific OpenSearch domain:

Reviewer 2: @harshavamsi Can you please address? I can approve once you have. Thanks!

> Be sure to update the access control to an IAM role/user. Do not use master user. In the following image, the IAM role used allows access to this specific OpenSearch domain:

Author: @natebower I don't see too much of a difference between those two interpretations. Either looks good to me.

      // Any other method to acquire a new Credentials object can be used.
      const credentialsProvider = defaultProvider();
      return credentialsProvider();
    },
  }),
  node: "https://search-xxx.region.es.amazonaws.com", // OpenSearch domain URL
});

```
### Creating a client connection in Ruby

The [opensearch-aws-sigv4](https://github.com/opensearch-project/opensearch-ruby/tree/main/opensearch-aws-sigv4) gem provides the `OpenSearch::Aws::Sigv4Client` class, which has all features of `OpenSearch::Client`. The only difference between these two clients is that `OpenSearch::Aws::Sigv4Client` requires an instance of `Aws::Sigv4::Signer` during instantiation to authenticate with AWS:

```ruby
require 'opensearch-aws-sigv4'
require 'aws-sigv4'
signer = Aws::Sigv4::Signer.new(service: 'es',
                                region: 'us-west-2',
                                access_key_id: 'key_id',
                                secret_access_key: 'secret')
client = OpenSearch::Aws::Sigv4Client.new({ log: true }, signer)
client.cluster.health
client.transport.reload_connections!
client.search q: 'test'
```

### Creating a client connection in .NET

> All required request signing is handled by the `AwsSigV4HttpConnection` implementation. By default, `AwsSigV4HttpConnection` uses the .NET AWS SDK's default credentials provider to acquire credentials from the environment. However, you may opt to pass in your own credentials provider, for example, to assume a role. Refer to the [OpenSearch.Net User Guide](https://github.com/opensearch-project/opensearch-net/blob/main/USER_GUIDE.md#opensearchnetauthawssigv4) for complete getting started instructions.

Reviewer: OpenSearch.Net is the name of the client, so no space needed.

var connection = new AwsSigV4HttpConnection();
var config = new ConnectionSettings(endpoint, connection);
var client = new OpenSearchClient(config);
```

### Creating a client connection in Rust

Request signing is configured using the [`Credentials::AwsSigV4`](https://docs.rs/opensearch/latest/opensearch/auth/enum.Credentials.html#variant.AwsSigV4) enum variant or its helper conversion from an AWS SDK config. See [aws-config](https://docs.rs/aws-config/latest/aws_config/) for other AWS credentials provider implementations, for example, to assume a role.

```rust
use opensearch::{
    cat::CatIndicesParts,
    http::transport::{SingleNodeConnectionPool, TransportBuilder},
    OpenSearch,
};
use url::Url;

let creds = aws_config::load_from_env().await;

let host = "https://search-xxx.region.es.amazonaws.com";
let transport = TransportBuilder::new(SingleNodeConnectionPool::new(Url::parse(host).unwrap()))
    .auth(creds.try_into()?)
    // .auth(Credentials::AwsSigV4(creds.credentials().unwrap().clone(), creds.region().unwrap().clone()))
    .build()?;
let client = OpenSearch::new(transport);
```

### Creating a client connection in PHP

> The PHP client uses the `setSigV4CredentialProvider` attribute to assume credentials from the the local credential store. Use the `setSigV4Region` attribute to set the region. 

Reviewer: @harshavamsi could you confirm

> The PHP client uses the `setSigV4CredentialProvider` attribute to assume credentials from the the local credential store. Use the `setSigV4Region` attribute to set the region. 

Author: This is AWS region. I can update this.

---
layout: post
title:  "AWS SigV4 support for OpenSearch clients"
authors:
  - vachshah
  - dblock
  - hvamsi
  - theotr
  - xtansia
  - mnkugler
  - mtimmerm
  - ssayakci

> date: 2022-12-09

Reviewer 3: @harshavamsi  Can you do a quick update to today's date? 2022-12-12 
thanks
